runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsResponseInterceptor.kt

```diff
@@ -33,57 +34,72 @@ internal val CHECKSUM_HEADER_VALIDATION_PRIORITY_LIST: List<String> = listOf(
  */
-
 @InternalApi
-public class FlexibleChecksumsResponseInterceptor<I>(
-    private val shouldValidateResponseChecksumInitializer: (input: I) -> Boolean,
+public class FlexibleChecksumsResponseInterceptor(
+    private val responseValidationRequired: Boolean,
+    private val responseChecksumValidation: HttpChecksumConfigOption?,
 ) : HttpInterceptor {
-
-    private var shouldValidateResponseChecksum: Boolean = false
-
     @InternalApi
     public companion object {
         // The name of the checksum header which was validated. If `null`, validation was not performed.
         public val ChecksumHeaderValidated: AttributeKey<String> = AttributeKey("ChecksumHeaderValidated")
     }
 
-    override fun readBeforeSerialization(context: RequestInterceptorContext<Any>) {
-        @Suppress("UNCHECKED_CAST")
-        val input = context.request as I
-        shouldValidateResponseChecksum = shouldValidateResponseChecksumInitializer(input)
-    }
-
     override suspend fun modifyBeforeDeserialization(context: ProtocolResponseInterceptorContext<Any, HttpRequest, HttpResponse>): HttpResponse {
-        if (!shouldValidateResponseChecksum) {
-            return context.protocolResponse
-        }
+        val logger = coroutineContext.logger<FlexibleChecksumsResponseInterceptor>()
 
-        val logger = coroutineContext.logger<FlexibleChecksumsResponseInterceptor<I>>()
+        val forcedToVerifyChecksum = responseValidationRequired || responseChecksumValidation == HttpChecksumConfigOption.WHEN_SUPPORTED
+        if (!forcedToVerifyChecksum) return context.protocolResponse
 
         val checksumHeader = CHECKSUM_HEADER_VALIDATION_PRIORITY_LIST
             .firstOrNull { context.protocolResponse.headers.contains(it) } ?: run {
             logger.warn { "User requested checksum validation, but the response headers did not contain any valid checksums" }
             return context.protocolResponse
         }
+        val serviceChecksumValue = context.protocolResponse.headers[checksumHeader]!!
+
+        if (serviceChecksumValue.isCompositeChecksum()) {
+            logger.debug { "Service returned composite checksum. Skipping validation." }
+            return context.protocolResponse
+        }
 
-        // let the user know which checksum will be validated
         logger.debug { "Validating checksum from $checksumHeader" }
         context.executionContext[ChecksumHeaderValidated] = checksumHeader
 
-        val checksumAlgorithm = checksumHeader.removePrefix("x-amz-checksum-").toHashFunction() ?: throw ClientException("could not parse checksum algorithm from header $checksumHeader")
+        val checksumAlgorithm = checksumHeader
+            .removePrefix("x-amz-checksum-")
+            .toHashFunction() ?: throw ClientException("Could not parse checksum algorithm from header $checksumHeader")
+
+        if (context.protocolResponse.body is HttpBody.Bytes) {
```

Comment:
`toHashingBody` doesn't support `Bytes` bodies. I decided to calculate the checksum in memory instead of adding support for `Bytes` to `toHashingBody`. It should be safe if the request is not being streamed but I think you're right that we're not following the spec

Reply:
> `toHashingBody` doesn't support `Bytes` bodies
Does it need to? The previous implementation never seemed to need it

Reply:
I think it does, the spec has some unit tests with non-streaming bodies

Reply:
Discussed offline and we'll be removing support for `HttpBody.Bytes` from `FlexibleChecksumsRequestInterceptor` because HTTP bodies are never bytes. [Source](https://github.com/smithy-lang/smithy-kotlin/blob/main/runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/OkHttpUtils.kt#L104-L120)

Reply:
I think removing support for `HttpBody.Bytes` is a bad idea. Our _current_ implementations of OkHttp and CRT engines do not return bytes, that's true, but future implementations or other engines wrapped by users _may_. In addition, interceptors may be used to rewrite response bodies by streaming them into memory, processing them, and then rewriting the body in a new form (possibly `HttpBody.Bytes`). I think we need to ensure each type of `HttpBody` is handled correctly.